# 12.04.2024

# my idea is to visualize the following 
# psuedo-chromosomes of unphased Sikem haploid assembly, the reference
# distribution of gaps, 
# distribution of snps phased by Hi-C,
# compare phase generated by PhaseGrass with Sikem unphased
# the chromosome-level phase blocks with SNP density
# the small phase blocks with SNP density


setwd("P:/Yutangchen/Sikem_hifi/plot_phase")

# import packages
library(stringr)
library(ggplot2)
library(ggnewscale)
# library(ggh4x)

# import data
# phase generated by PhaseGrass pipeline with short reads, long reads, Hi-C
wgs_short_phase <- read.table("phase_check_hifi_vs_ont.txt", header = T, stringsAsFactors = F, sep = '\t')
# Hi-C phased SNPs, generated by hapcut2, only the largest phase block
hic_phased_snp <- read.table('HiC_phased_SNP.txt', header = T, stringsAsFactors = F, sep = '\t')
# Sikem largest blocks
Sikem_large <- read.table('Sikem_largest_block.txt', header = T, stringsAsFactors = F, sep = '\t')
# Sikem small blocks
Sikem_small <- read.table('Sikem_small_block.txt', header = T, stringsAsFactors = F, sep = '\t')
# chromosome length of Sikem unphased
Sikem_fai <- read.table('sikem.hifiasm.juicer.fasta.chr.fa.fai', header = F, stringsAsFactors = F)
# gaps in chromosomes of Sikem unphased
Sikem_gap <- read.table('sikem.hifiasm.juicer.fasta.chr.gap.bed', header = F, stringsAsFactors = F)

# get gap
Sikem_gap <- Sikem_gap[, 1:3]
colnames(Sikem_gap) <- c('chr', 'start', 'end')
Sikem_gap$chr <- str_to_title(Sikem_gap$chr)

# make the chr capital in fai
Sikem_fai <- Sikem_fai[, c(1:2)]
colnames(Sikem_fai) <- c('chr', 'length')
Sikem_fai$chr <- str_to_title(Sikem_fai$chr)

# get hifi largest blocks
Sikem_large_hifi <- Sikem_large[Sikem_large$seqtech == 'HiFi', ]
# get hifi small blocks
Sikem_small_hifi <- Sikem_small[Sikem_small$seqtech == 'HiFi', ]

# get ont largest blocks
Sikem_large_ont <- Sikem_large[Sikem_large$seqtech == 'ONT', ]
# get ont small blocks
Sikem_small_ont <- Sikem_small[Sikem_small$seqtech == 'ONT', ]

##############################################################################################################################
# visualize the data
ggplot(data = Sikem_fai, aes(x = length, y = length)) +
  
  # determine x and y scale
  scale_x_continuous(limits = c(0, 430), expand = c(0.01, 0.01)) +
  scale_y_continuous(breaks = c(1, 2.5, 4, 5.5, 7, 8.5, 10, 11.5, 13),
                     limits = c(0, 14),
                     # labels = c('Pseudo-chromosome',
                     #            'Hi-C phased SNPs',
                     #            'ONT chromosome-level phase block',
                     #            'ONT small phase blocks',
                     #            'HiFi chromosome-level phase block',
                     #            'HiFi small phase blocks',
                     #            'DipGrass ONT vs reference',
                     #            'DipGrass HiFi vs reference',
                     #            'DipGrass HiFi vs DipGrass ONT')
                     labels = c('1', '2', '3', '4', '5', '6', '7', '8', '9')) +
  
  # plot all chromosomes in one rwo, each column represents one chr
  # facet_grid(cols = vars(chr), scales = 'free') +
  facet_grid(rows = vars(chr), scales = 'free') +
  
  # first visualize chromosomes using rectangles
  geom_rect(data = Sikem_fai, aes(xmin = 0, ymin = 0.5, xmax = length/10^6, ymax = 1.5), fill = 'gray80', color = 'black', inherit.aes = F) + 
  
  # now add gaps to chromosomes as rectangles
  geom_rect(data = Sikem_gap, aes(xmin = start/10^6, ymin = 0.5, xmax = end/10^6, ymax = 1.5), fill = 'black', color = 'black', inherit.aes = F) +
  
  # now hic snp density through chromosomes
  geom_rect(data = hic_phased_snp, aes(xmin = start/10^6, ymin = 2, xmax = end/10^6, ymax = 3, fill = snp_number), inherit.aes = F) + 
  scale_fill_gradient(low = 'white', high = 'orange', name = '# of SNPs per 1 Mb') + 
  
  # new scale color
  new_scale_fill() +
  
  # now the ont largest phase block 
  geom_rect(data = Sikem_large_ont, aes(xmin = start/10^6, ymin = 3.5, xmax = end/10^6, ymax = 4.5, fill = nsnp), inherit.aes = F) +
  scale_fill_gradient(low = 'white', high = 'steelblue1', name = '# of SNPs per 1 Mb') + 
  
  # new scale color
  new_scale_fill() +
  
  # now the ont small phase blocks
  geom_point(data = Sikem_small_ont, aes(x = block/10^6, y = 5.5, fill = variants), size = 2, color = 'black', shape = 21, position = position_jitter(width = 0.3, height = 0.3)) +
  scale_fill_gradient(low = 'white', high = 'blue', name = '# of SNPs in ONT small blocks') +
  
  # new scale color
  new_scale_fill() +
  
  # now the hifi largest phase block 
  geom_rect(data = Sikem_large_hifi, aes(xmin = start/10^6, ymin = 6.5, xmax = end/10^6, ymax = 7.5, fill = nsnp), inherit.aes = F) +
  scale_fill_gradient(low = 'white', high = 'deeppink1', name = '# of SNPs per 1 Mb') +
  
  # new scale color
  new_scale_fill() +
  
  # now the ont small phase blocks
  geom_point(data = Sikem_small_hifi, aes(x = block/10^6, y = 8.5, fill = variants), size = 2, color = 'black', shape = 21, position = position_jitter(width = 0.3, height = 0.3)) +
  scale_fill_gradient(low = 'white', high = 'red', name = '# of SNPs in HiFi small blocks') +
  
  # new scale color
  new_scale_fill() +
  
  # now phasegras ont vs unphased sikem
  geom_rect(data = wgs_short_phase, aes(xmin = start/10^6, ymin = 9.5, xmax = end/10^6, ymax = 10.5, fill = ontvsref*100), inherit.aes = F) +
  scale_fill_gradient2(low = "blue", mid = 'gray', high = "red", midpoint = 50, name = '% of non-reference allele per 1 Mb') +
  
  # now phasegras hifi vs unphased sikem
  geom_rect(data = wgs_short_phase, aes(xmin = start/10^6, ymin = 11, xmax = end/10^6, ymax = 12, fill = hifivsref*100), inherit.aes = F) +
  scale_fill_gradient2(low = "blue", mid = 'gray', high = "red", midpoint = 50, name = '% of non-reference allele per 1 Mb') +
  
  # new scale color
  new_scale_fill() +
  
  # now hifi vs ont
  geom_rect(data = wgs_short_phase, aes(xmin = start/10^6, ymin = 12.5, xmax = end/10^6, ymax = 13.5, fill = hifivsont*100), inherit.aes = F) +
  scale_fill_gradient(low = 'white', high = 'darkgreen', name = '% of phase differences per 1 Mb') +
  
  # axis lables
  xlab('Base pair position in Sikem unphased psuedo-chromosomes (Mb)') +
  ylab('') +
  
  # theme
  theme(strip.text.y = element_text(face = "plain", size = 20),
        # strip.background = element_blank(),
        legend.position="bottom",
        legend.justification.bottom = "right",
        legend.text = element_text(size=10),
        legend.title = element_text(size = 15,  hjust = 0.5),
        legend.title.position = 'top',
        legend.key = element_rect(fill = NA, colour = NA),
        legend.key.height = unit(0.3, 'cm'), 
        legend.key.width = unit(0.8, 'cm'), 
        panel.border = element_blank(),
        panel.background = element_blank(),
        # panel.background = element_rect(fill = "white", colour = "black", linewidth = 0.5, linetype = "solid"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.spacing = unit(0.8, "lines"),
        axis.title.x = element_text(color="black", size=20, face="plain"),
        axis.title.y = element_blank(),
        axis.text.x = element_text(size = 15),
        axis.text.y = element_text(size = 15),
        axis.line.x = element_line(color = 'black'))

ggsave('Sikem_phase_evaluation.pdf', device = 'pdf', width = 20, height = 15)

