This folder contains scripts that were used to evaluate the phasing quality of Sikem

plot_phase_sikem.R visualizes the following:

1 psuedo-chromosomes of unphased Sikem haploid assembly, the reference

2 distribution of gaps in the reference 

3 distribution of snps phased by Hi-C,

4 compare phase generated by PhaseGrass (for both combinations, ont-hic and pacbio hifi-hic) with Sikem unphased

5 the chromosome-level phase blocks with SNP density (for both combinations, ont-hic and pacbio hifi-hic)

6 the small phase blocks with SNP density (for both combinations, ont-hic and pacbio hifi-hic)

7 show the number of different alleles between ont-hic phasing and pacbio hifi-hic phasing

Input data for 1 and 2 were obtained with sikem unphased haploid assembly. Chromosome size was genreated with samtools faidx, and the gap position was found with seqkit locate -p N+ -G -P -r -i --bed {input} > {output}

Input data for 3 were generated with check_hic_SNP.R. This R scripts takes the input generated by get_largest_phase_block.sh. get_largest_phase_block.sh takes the phased VCF genreated by HapCUT2 with Hi-C data.

Input data for 4 and 7 were genreated with phase_hifi_vs_ont.R. 
First, Phased SNPs extracted by get_phase.sh from the phased VCF genrated by WhatsHap. 
Second, The ratio of non-reference allele per 1 Mb window was calculated with check_phase_with_snp_density.R.
Last, output file generated by check_phase_with_snp_density.R was fed to phase_hifi_vs_ont.R.

Input data for 5 and 6 were generated with DH647_phase_block_new.R (yes, I uesed this same script used for DH647, just changed the input files). This R script takes the phase block list generatd by WhatsHap and the phased SNPs (in the chromosome-level and small phase blocks separately) as input. The phased SNPs with their associated phase blocks were extracted from the phased VCF genreated by WhatsHap using get_phase_block.sh

Dataflow
![phasing_quality_evaluation_Sikem](https://github.com/user-attachments/assets/63fbdb4c-b029-40a1-86a4-7cf4717b97cd)
