Comparing haplotypes of Sikem generated by PhaseGrass with haplotypes from a genetic linkage map of Sikem

The WHATIDID.txt recorded what I have done

The Snakefile is the GBS read mapping pipeline and SNP calling pipeline

However, to have phased alleles in the genetic linakge map, the SNP calling pipeline suggested by lepmap3 should be used, which is in lepmap3_variant_calling.sh

Folder lepmap contains the scripts I used to generate the genetic linkage map

make_pedigree.R generates the pedigree file required by lepmap3

cal_distance.sh employs lepmap3 to calculate the genetic linkage map disdance based on the phasical order of the SNPs in the haplome generated by PhaseGrass.
cal_distance.sh needs map2genotypes_phased.awk from lepmap3, and bedtools to finish the whole script.

The comparision between haplotypes generated by PhaseGrass and lepmap3 was visualized with physcial_map_vs_genetic_linkage_map.R
